go/teams/transactions.go

```diff
@@ -0,0 +1,495 @@
+		if chainUv.Uid == uid && team.chain().getUserRole(chainUv) != keybase1.TeamRole_NONE {
+			tx.RemoveMember(chainUv)
+		}
+	}
+}
+
+func (tx *AddMemberTx) SweepKeybaseInvites(uid keybase1.UID) {
+	team := tx.team
+	for _, invite := range team.chain().inner.ActiveInvites {
+		if inviteUv, err := invite.KeybaseUserVersion(); err == nil {
+			if inviteUv.Uid == uid {
+				tx.CancelInvite(invite.Id)
+			}
+		}
+	}
+}
+
+// High level API:
+
+// AddMemberTransaction will add member by username and role. It
+// checks if given username can become crypto member or a PUKless
+// member. It will also clean up old invites and memberships if
+// necessary.
+func (tx *AddMemberTx) AddMemberTransaction(ctx context.Context, username string, role keybase1.TeamRole) error {
+	team := tx.team
+	g := team.G()
+
+	g.Log.CDebugf(ctx, "AddMemberTransaction(%s, %v) to team %q", username, role, team.Name())
+
+	inviteRequired := false
+	normalizedUsername, uv, err := loadUserVersionPlusByUsername(ctx, g, username)
+	g.Log.CDebugf(ctx, "AddMemberTransaction: loaded user %q -> (%q, %v, %v)", username, normalizedUsername, uv, err)
+	if err != nil {
+		if err == errInviteRequired {
+			inviteRequired = true
+			g.Log.CDebugf(ctx, "Invite required for %v", uv)
+		} else {
+			return err
+		}
+	}
+
+	// Do not do partial updates here. If error is returned, it is
+	// assumed that tx is untouched, and caller can continue with
+	// other attempts. This is used in batch member adds, when even if
+	// some users can't be added, it skips them and continues with
+	// others.
+
+	if role == keybase1.TeamRole_OWNER && team.IsSubteam() {
+		return NewSubteamOwnersError()
+	}
+
+	if team.IsMember(ctx, uv) {
+		if inviteRequired {
+			return fmt.Errorf("user is already member but we got errInviteRequired")
+		}
+		return libkb.ExistsError{Msg: fmt.Sprintf("user %s is already a member of team %q",
+			normalizedUsername, team.Name())}
+	}
+
+	existingUV, err := team.UserVersionByUID(ctx, uv.Uid)
+	if err == nil {
+		// TODO: Might be able to collapse the two assertions together - the one
+		// above with team.IsMember and this one which checking Uid/Eldest.
+
+		// There is an edge case where user is in the middle of
+		// resetting (after reset, before provisioning) and has
+		// EldestSeqno=0.
+		if !inviteRequired && existingUV.EldestSeqno > uv.EldestSeqno {
+			return fmt.Errorf("newer version of user %s (uid:%s) already exists in the team %q (%v > %v)",
+				normalizedUsername, uv.Uid, team.Name(), existingUV.EldestSeqno, uv.EldestSeqno)
+		}
+	}
+
+	curInvite, err := team.chain().FindActiveInvite(uv.TeamInviteName(), keybase1.NewTeamInviteTypeDefault(keybase1.TeamInviteCategory_KEYBASE))
+	if err != nil {
+		if _, ok := err.(libkb.NotFoundError); !ok {
+			return err
+		}
+		curInvite = nil
+	}
+	if curInvite != nil && inviteRequired {
+		return libkb.ExistsError{Msg: fmt.Sprintf("user %s is already a pukless member of team %q",
+			normalizedUsername, team.Name())}
```

Comment:
This error message seems like something users might actually see, so how about replacing "a pukless member of" with "invited to"